packages/test/test-end-to-end-tests/src/test/dataMigration.spec.ts

```diff
@@ -0,0 +1,425 @@
+		submitSummaryFn: (_summaryOp, _referenceSequenceNumber?) => 0,
+		submitBatchFn: (_batch, _referenceSequenceNumber?) => 0,
+		submitSignalFn: (_contents: any) => {},
+		disposeFn: context.disposeFn,
+		closeFn: context.closeFn,
+		updateDirtyContainerState: context.updateDirtyContainerState,
+		getAbsoluteUrl: context.getAbsoluteUrl,
+		getSpecifiedCodeDetails: () => {
+			return { package: codeDetailsVersion };
+		},
+		getLoadedFromVersion: context.getLoadedFromVersion,
+		attachState: context.attachState,
+	};
+	return readonlyContext;
+}
+
+class ContainerRuntimeFactoryWithDataMigration extends RuntimeFactoryHelper {
+	private _currentRuntimeFactory?: ContainerRuntimeFactoryWithDefaultDataStore;
+	private oldRuntime?: ContainerRuntime;
+	private get currentRuntimeFactory(): ContainerRuntimeFactoryWithDefaultDataStore {
+		assert(
+			this._currentRuntimeFactory !== undefined,
+			"current should have been set before getting!",
+		);
+		return this._currentRuntimeFactory;
+	}
+
+	private clientDetailsType?: string;
+
+	constructor(
+		private readonly v1: ContainerRuntimeFactoryWithDefaultDataStore,
+		private readonly v1ToV2: ContainerRuntimeFactoryWithDefaultDataStore,
+		private readonly v2: ContainerRuntimeFactoryWithDefaultDataStore,
+		private readonly deferred: Deferred<{
+			summarizer: ContainerRuntime;
+			readonlyRuntime: ContainerRuntime;
+		}>,
+	) {
+		super();
+	}
+	public async preInitialize(
+		context: IContainerContext,
+		existing: boolean,
+	): Promise<ContainerRuntime> {
+		this.clientDetailsType = context.clientDetails.type;
+		const codeDetails = context.getSpecifiedCodeDetails
+			? context.getSpecifiedCodeDetails()
+			: undefined;
+		assert(codeDetails !== undefined, "get code details failed");
+
+		this._currentRuntimeFactory = codeDetails.package === "v1" ? this.v1 : this.v2;
+		if (this.clientDetailsType === detachedClientType) {
+			const detachedContext = createReadOnlyContext(context, "v2");
+			this._currentRuntimeFactory = this.v1ToV2;
+			return this.currentRuntimeFactory.preInitialize(detachedContext, existing);
+		}
+
+		return this.currentRuntimeFactory.preInitialize(context, existing);
+	}
+
+	public async instantiateFirstTime(runtime: ContainerRuntime): Promise<void> {
+		return this.currentRuntimeFactory.instantiateFirstTime(runtime);
+	}
+
+	public async instantiateFromExisting(runtime: ContainerRuntime): Promise<void> {
+		return this.currentRuntimeFactory.instantiateFromExisting(runtime);
+	}
+
+	public async hasInitialized(runtime: ContainerRuntime): Promise<void> {
+		await this.currentRuntimeFactory.hasInitialized(runtime);
+
+		if (this.clientDetailsType === summarizerClientType) {
+			this.oldRuntime = runtime;
+			const readonlyRuntime = await runtime.loadDetachedAndTransitionFn();
+			this.deferred.resolve({ summarizer: this.oldRuntime, readonlyRuntime });
+		}
+	}
+}
+
+class SharedCellToSharedMapFactory implements IChannelFactory {
```

Comment:
If I got right, this is essentially on-the-fly transformer for node in the graph (dds), if using this kind of language.

Reply:
Yep